Add GPU support for ONNXRuntime

hqucms opened the pull request:
#### PR description:

This is a technical PR that allows the GPU support of ONNXRuntime (https://github.com/cms-sw/cmsdist/pull/6776) to be easily enabled in CMSSW.

Example usage:

```c++
  using namespace cms::Ort;
  std::string model_path = edm::FileInPath("PhysicsTools/ONNXRuntime/test/data/model.onnx").fullPath();
  auto session_options = ONNXRuntime::defaultSessionOptions(/*use_cuda=*/true);
  ONNXRuntime rt(model_path, &session_options);
```

#### PR validation:

Unit tests are updated to cover all three options.

hqucms committed:
Add GPU support for ONNXRuntime.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36963/28320

- This PR adds an extra 20KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36963/28320/code-checks.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36963/28320/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36963/28323

- This PR adds an extra 20KB to repository

cmsbuild commented:
A new Pull Request was created by  @hqucms (Huilin Qu) for master.

It involves the following packages:

- PhysicsTools/ONNXRuntime (**reconstruction**)


@jpata, @cmsbuild, @clacaputo, @slava77 can you please review it and eventually sign? Thanks.
@riga this is something you requested to watch as well.
@perrotta, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

tvami commented:
@cmsbuild , please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22413/summary.html
**COMMIT**: 3ef28ffabbc6223845e44ff259e7bc7871287492
**CMSSW**: CMSSW_12_3_X_2022-02-14-1100/slc7_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/36963/22413/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 5 differences found in the comparisons
 - DQMHistoTests: Total files compared: 46
 - DQMHistoTests: Total histograms compared: 3764435
 - DQMHistoTests: Total failures: 13
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 3764399
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.004 KiB( 45 files compared)
 - DQMHistoSizes: changed ( 312.0 ): 0.004 KiB        MessageLogger/Warnings
 - Checked 193 log files, 42 edm output root files, 46 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_3_X_2022-02-14-1100+a42efb/48371/triggerResults): no differences found

cmsbuild commented:
New categories assigned: heterogeneous

@fwyzard,@makortel you have been requested to review this Pull request/Issue and eventually sign? Thanks

fwyzard commented on PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc:
```diff
@@ -78,6 +77,35 @@ namespace cms::Ort {
+        if (gpu_mode == force_gpu) {
+          throw cms::Exception("RuntimeError") << "No GPU detected, cannot run ONNXRuntime on GPU.";
+        } else {
+          std::cout << "[ONNXRuntime] No GPU detected, will run on CPU." << std::endl;
```
please don't use `std::cout`, as it cannot be silenced from the job configuration

hqucms replied:
Would `edm::LogInfo` work?

fwyzard replied:
Sure, or maybe `edm::LogWarning` if it's supposed to be a more visible message.

hqucms commented:
> * how complicated would it be to extend to the Async mode in CMSSW?

I have never tried to use the async mode in CMSSW -- can you point me to any documentation or example?

> * now it grabs the whole GPU, how difficult would it be to extend to use streams?

It does not grabs the whole GPU, so in principle it can be used in multiple streams simultaneously I think (to be tested of course).

fwyzard commented on PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc:
```diff
@@ -104,6 +132,10 @@ namespace cms::Ort {
       } else {
         input_dims = input_shapes[input_pos];
         // rely on the given input_shapes to set the batch size
+        if (input_dims[0] != batch_size) {
+          throw cms::Exception("RuntimeError") << "The first element of `input_shapes` (" << input_dims[0]
+                                               << ") does not match the given `batch_size` (" << batch_size << ")";
+        }
```
this change is unrelated to the use of GPUs, isn't it ?

hqucms replied:
No it's unrelated to GPU -- just a helpful error message for users.

fwyzard commented:
What NVIDIA GPU architectures (SM 6.0, 7.0, 7.5, _etc._) does ONNX support ?
Is it tied to how we use CUDA in CMSSW, or is it independent from it ?

fwyzard commented:
> > * how complicated would it be to extend to the Async mode in CMSSW?
> 
> I have never tried to use the async mode in CMSSW -- can you point me to any documentation or example?

It looks like most (all ?) CMSSW modules that use ONNX are already `stream` modules and share the same `ONNXRuntime` instance,_e.g._:
```c++
class BoostedJetONNXJetTagsProducer : public edm::stream::EDProducer<edm::GlobalCache<ONNXRuntime>> {
  ...
```

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36963/28338

- This PR adds an extra 20KB to repository

cmsbuild commented:
Pull request #36963 was updated. @makortel, @slava77, @clacaputo, @cmsbuild, @fwyzard, @jpata can you please check and sign again.

makortel commented:
Does ONNX itself provide an asynchronous API for inference?

makortel commented:
In addition to what Andrea wrote in https://github.com/cms-sw/cmssw/pull/36963#issuecomment-1040314622, the documentation of `ExternalWork` extension is documented in
https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkStreamModuleInterface#edm_ExternalWork

fwyzard commented:
The main concern I have is about the interface. Figuring out something more flixible now could avoid making changes later, but it could be that's easier to take the changes like this, and improve on the interface later on, once we have more experience with ONNX on GPUs.

Anyway
  - we shouldn't use "GPU" to mean "only an NVIDIA GPU"; maybe use "CUDA" instead ?
  - the choice to use of GPUs (yes, no, auto) should be made consistently with the rest of CMSSW, as in #36699; @makortel what do you think ?
  - it might actually be useful to expose the setting about the `options.cuda_mem_limit`, to make it easier to share the GPU with other modules in the same job, but also other jobs running on the same machine
  - does `device_id = 0` tell ONXX to only use the first available GPU ? this is probably OK to start with, but later on we need to figure out how to deal with multiple GPUs, and how pass this information to the ONNX runtime from the framework

makortel commented on PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc:
```diff
@@ -78,6 +78,35 @@ namespace cms::Ort {
+      bool is_gpu_available = (status == cudaSuccess && devices > 0);
+      if (is_gpu_available) {
+        // https://www.onnxruntime.ai/docs/reference/execution-providers/CUDA-ExecutionProvider.html
+        OrtCUDAProviderOptions options;
```
I see in https://github.com/cms-sw/cmsdist/blob/f2fd150c7f36845f4f5a00a60b24dd33d265ec7b/onnxruntime.spec#L12-L14 and https://github.com/cms-sw/cmsdist/blob/f2fd150c7f36845f4f5a00a60b24dd33d265ec7b/onnxruntime.spec#L23-L27
```spec
%if "%{cmsos}" != "slc7_aarch64"
Requires: cuda cudnn
%endif
...
%if "%{cmsos}" != "slc7_aarch64"
USE_CUDA=ON
%else
USE_CUDA=OFF
%endif
```
that the CUDA backend is not supported for all of our platforms (even if slc7 is going to go away soonish(?)). What happens with these options if CUDA has been disabled in ONNX? Like, are the classes available at compile time?

makortel replied:
Given the condition `Requires: cuda` it would be better to make this package to explicitly depend on `cuda` external.

hqucms replied:
> that the CUDA backend is not supported for all of our platforms (even if slc7 is going to go away soonish(?)). 

The restriction here should be `aarch64` and not really `slc7` I think:)

> What happens with these options if CUDA has been disabled in ONNX? Like, are the classes available at compile time?

The classes should be available at compile time but I guess the `AppendExecutionProvider_CUDA` call will fail at run time.

makortel replied:
> > that the CUDA backend is not supported for all of our platforms (even if slc7 is going to go away soonish(?)).
> 
> The restriction here should be `aarch64` and not really `slc7` I think:)

Thanks, then we should fix that :) Let me tag @smuzaffar here too.

> > What happens with these options if CUDA has been disabled in ONNX? Like, are the classes available at compile time?
> 
> The classes should be available at compile time but I guess the `AppendExecutionProvider_CUDA` call will fail at run time.

Ok, then this code could be fine as a first attempt (I'd expect it would need to be changed at some point, but hard to guess now how), assuming `PhysicsTools/ONNXRuntime` would gain explicit dependence on `cuda` external.

smuzaffar replied:
@hqucms @makortel , problem is with cudnn distribution for aarch64 for slc7. We use pre-build cudnn distributiuon ( https://github.com/cms-sw/cms-bot/blob/master/pr_testing/run-pr-comparisons#L124 ) . This was built for newer kernel ( and works on cs8) but failed for run/package for slc7. That is why the `cuda cudnn` is not included for `slc7_aarch64`.

makortel commented on PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc:
```diff
@@ -78,6 +78,35 @@ namespace cms::Ort {
+    if (gpu_mode != no_gpu) {
+      // detect if there is GPU
+      int devices = 0;
+      auto status = cudaGetDeviceCount(&devices);
```
I wonder if `CUDAService` should be used here instead for figuring our whether CUDA can be used or not?

hqucms replied:
It should be used in the `EDModule` or directly here?

makortel replied:
`CUDAService` can be used directly here. (this function is called from the EDModule' s, right? e.g. from their constructor)

makortel commented on PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc:
```diff
@@ -78,6 +78,35 @@ namespace cms::Ort {
+      } else {
+        // if GPU is not available
+        if (gpu_mode == force_gpu) {
+          throw cms::Exception("RuntimeError") << "No GPU detected, cannot run ONNXRuntime on GPU.";
```
This would be better as
```suggestion
          throw cms::Exception("UnavailableAccelerator") << "No GPU detected, cannot run ONNXRuntime on GPU.";
```
to align with #36699 (the string category should get mapped to `edm::errors::UnavailableAccelerator`) when #36699 gets in).

makortel reviewed (commented):
Question beyond this PR: I'm wondering how ONNX supports (or plans to support) AMD and Intel GPUs? Are they going to bundle the support for all into a single dynamic library (at least what the "frontend" is concerned)? Or would there be one `onnxruntime` build for each GPU vendor? Or something else?

fwyzard commented:
enable gpu

fwyzard commented:
please test

tvami commented:
@cmsbuild please test with https://github.com/cms-sw/cmssw/pull/36972

hqucms commented:
> What NVIDIA GPU architectures (SM 6.0, 7.0, 7.5, _etc._) does ONNX support ?

I think it's listed [here](https://github.com/cms-externals/onnxruntime/blob/cms/v1.7.2/cmake/CMakeLists.txt#L1190-L1203) which seems to go back to even SM 3.7.

> Is it tied to how we use CUDA in CMSSW, or is it independent from it ?

It's based on the CUDA and cuDNN libraries in CMSSW.

cmsbuild commented:
-1

**Failed Tests**: UnitTests RelVals RelVals-INPUT AddOn 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/summary.html
**COMMIT**: e2a6cf6997a7986d66060059dde2136c1cb5be9b
**CMSSW**: CMSSW_12_3_X_2022-02-14-2300/slc7_amd64_gcc10
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/36963/22439/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - @pmandrik cms-sw/cmssw#36958
 - @bsunanda cms-sw/cmssw#36945
 - @mmusich cms-sw/cmssw#36962
 - @lecriste cms-sw/cmssw#36712
 - @fabferro cms-sw/cmssw#36960
 - @dan131riley cms-sw/cmssw#36965
 - @tvami cms-sw/cmssw#36942
 - @andrea21z cms-sw/cmssw#36179
 - @kakwok cms-sw/cmssw#36929
 - @dinyar cms-sw/cmssw#36928
 - @bsunanda cms-sw/cmssw#36964
 - @wouf cms-sw/cmssw#36918
 - @missirol cms-sw/cmssw#36951
 - @thomreis cms-sw/cmssw#36935

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/git-merge-result


## Unit Tests

I found errors in the following unit tests:

<pre>---> test runtestPhysicsToolsPatAlgos had ERRORS
---> test runtestTqafTopEventProducers had ERRORS
---> test runtestTqafTopEventSelection had ERRORS
---> test runtestTqafTopHitFit had ERRORS
and more ...
</pre>

## RelVals

- [136.7611](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/136.7611_RunJetHT2016E_reminiaod+RunJetHT2016E_reminiaod+REMINIAOD_data2016_HIPM+HARVESTDR2_REMINIAOD_data2016_HIPM)
```
----- Begin Fatal Exception 15-Feb-2022 18:19:29 CET-----------------------
An exception of category 'ConfigFileReadError' occurred while
   [0] Processing the python configuration file named step2_PAT_DQM.py
Exception Message:
 unknown python problem occurred.
RuntimeError: An exception of category 'FileInPathError' occurred.
Exception Message:
edm::FileInPath unable to find file RecoMuon/MuonIdentification/data/mvaID.onnx anywhere in the search path.
The search path is defined by: CMSSW_SEARCH_PATH
${CMSSW_SEARCH_PATH} is: /cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/36963/22439/CMSSW_12_3_X_2022-02-14-2300/poison:/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/36963/22439/CMSSW_12_3_X_2022-02-14-2300/src:/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/36963/22439/CMSSW_12_3_X_2022-02-14-2300/external/slc7_amd64_gcc10/data:/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/poison:/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/src:/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/external/slc7_amd64_gcc10/data
Current directory is: /data/cmsbld/jenkins/workspace/ib-run-pr-relvals/runTheMatrix-results/136.7611_RunJetHT2016E_reminiaod+RunJetHT2016E_reminiaod+REMINIAOD_data2016_HIPM+HARVESTDR2_REMINIAOD_data2016_HIPM


At:
  /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/python/FWCore/ParameterSet/Types.py(840): insertInto
  /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/python/FWCore/ParameterSet/Mixins.py(377): insertContentsInto
  /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/python/FWCore/ParameterSet/Mixins.py(512): insertInto
  /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/python/FWCore/ParameterSet/Modules.py(161): insertInto
  /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/python/FWCore/ParameterSet/Config.py(1145): _insertManyInto
  /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/python/FWCore/ParameterSet/Config.py(1400): fillProcessDesc
  <string>(2): <module>

----- End Fatal Exception -------------------------------------------------
```
- [136.8311](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/136.8311_RunJetHT2017F_reminiaod+RunJetHT2017F_reminiaod+REMINIAOD_data2017+HARVEST2017_REMINIAOD_data2017)
```
----- Begin Fatal Exception 15-Feb-2022 18:19:29 CET-----------------------
An exception of category 'ConfigFileReadError' occurred while
   [0] Processing the python configuration file named step2_PAT_DQM.py
Exception Message:
 unknown python problem occurred.
RuntimeError: An exception of category 'FileInPathError' occurred.
Exception Message:
edm::FileInPath unable to find file RecoMuon/MuonIdentification/data/mvaID.onnx anywhere in the search path.
The search path is defined by: CMSSW_SEARCH_PATH
${CMSSW_SEARCH_PATH} is: /cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/36963/22439/CMSSW_12_3_X_2022-02-14-2300/poison:/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/36963/22439/CMSSW_12_3_X_2022-02-14-2300/src:/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/36963/22439/CMSSW_12_3_X_2022-02-14-2300/external/slc7_amd64_gcc10/data:/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/poison:/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/src:/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/external/slc7_amd64_gcc10/data
Current directory is: /data/cmsbld/jenkins/workspace/ib-run-pr-relvals/runTheMatrix-results/136.8311_RunJetHT2017F_reminiaod+RunJetHT2017F_reminiaod+REMINIAOD_data2017+HARVEST2017_REMINIAOD_data2017


At:
  /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/python/FWCore/ParameterSet/Types.py(840): insertInto
  /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/python/FWCore/ParameterSet/Mixins.py(377): insertContentsInto
  /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/python/FWCore/ParameterSet/Mixins.py(512): insertInto
  /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/python/FWCore/ParameterSet/Modules.py(161): insertInto
  /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/python/FWCore/ParameterSet/Config.py(1145): _insertManyInto
  /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/python/FWCore/ParameterSet/Config.py(1400): fillProcessDesc
  <string>(2): <module>

----- End Fatal Exception -------------------------------------------------
```
- [136.88811](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/136.88811_RunJetHT2018D_reminiaodUL+RunJetHT2018D_reminiaodUL+REMINIAOD_data2018UL+HARVEST2018_REMINIAOD_data2018UL)
```
----- Begin Fatal Exception 15-Feb-2022 18:19:31 CET-----------------------
An exception of category 'ConfigFileReadError' occurred while
   [0] Processing the python configuration file named step2_PAT_DQM.py
Exception Message:
 unknown python problem occurred.
RuntimeError: An exception of category 'FileInPathError' occurred.
Exception Message:
edm::FileInPath unable to find file RecoMuon/MuonIdentification/data/mvaID.onnx anywhere in the search path.
The search path is defined by: CMSSW_SEARCH_PATH
${CMSSW_SEARCH_PATH} is: /cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/36963/22439/CMSSW_12_3_X_2022-02-14-2300/poison:/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/36963/22439/CMSSW_12_3_X_2022-02-14-2300/src:/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/36963/22439/CMSSW_12_3_X_2022-02-14-2300/external/slc7_amd64_gcc10/data:/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/poison:/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/src:/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/external/slc7_amd64_gcc10/data
Current directory is: /data/cmsbld/jenkins/workspace/ib-run-pr-relvals/runTheMatrix-results/136.88811_RunJetHT2018D_reminiaodUL+RunJetHT2018D_reminiaodUL+REMINIAOD_data2018UL+HARVEST2018_REMINIAOD_data2018UL


At:
  /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/python/FWCore/ParameterSet/Types.py(840): insertInto
  /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/python/FWCore/ParameterSet/Mixins.py(377): insertContentsInto
  /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/python/FWCore/ParameterSet/Mixins.py(512): insertInto
  /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/python/FWCore/ParameterSet/Modules.py(161): insertInto
  /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/python/FWCore/ParameterSet/Config.py(1145): _insertManyInto
  /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/python/FWCore/ParameterSet/Config.py(1400): fillProcessDesc
  <string>(2): <module>

----- End Fatal Exception -------------------------------------------------
```
<details>
<summary>Expand to see more relval errors ...</summary>

- [4.53](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/4.53_RunPhoton2012B+RunPhoton2012B+HLTD+RECODR1reHLT+HARVESTDR1reHLT)
- [136.731](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/136.731_RunSinglePh2016B+RunSinglePh2016B+HLTDR2_2016+RECODR2_2016reHLT_skimSinglePh_HIPM+HARVESTDR2_skimSinglePh)
- [136.793](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/136.793_RunDoubleEG2017C+RunDoubleEG2017C+HLTDR2_2017+RECODR2_2017reHLT_skimDoubleEG_Prompt+HARVEST2017_skimDoubleEG)
- [136.874](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/136.874_RunEGamma2018C+RunEGamma2018C+HLTDR2_2018+RECODR2_2018reHLT_skimEGamma_Offline_L1TEgDQM+HARVEST2018_L1TEgDQM)
- [135.4](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/135.4_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS)
- [158.01](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/158.01_HydjetQ_reminiaodPbPb2018_INPUT+HydjetQ_reminiaodPbPb2018_INPUT+REMINIAODHI2018PPRECOMB+HARVESTHI2018PPRECOMINIAOD)
- [139.001](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/139.001_RunMinimumBias2021+RunMinimumBias2021+HLTDR3_2021+RECODR3_MinBiasOffline+HARVESTD2021MB)
- [9.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/9.0_Higgs200ChargedTaus+Higgs200ChargedTaus+DIGI+RECO+HARVEST)
- [25.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/25.0_TTbar+TTbar+DIGI+RECOAlCaCalo+HARVEST+ALCATT)
- [1306.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/1306.0_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15)
- [2018.1](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/2018.1_TTbar_13_UP18+TTbarFS_13_UP18+HARVESTUP18FS+MINIAODMCUP18FS)
- [1330.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/1330.0_ZMM_13+ZMM_13+DIGIUP15+RECOUP15_L1TMuDQM+HARVESTUP15_L1TMuDQM+NANOUP15)
- [312.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/312.0_Pyquen_ZeemumuJets_pt10_2760GeV_2021+Pyquen_ZeemumuJets_pt10_2760GeV_2021+DIGIHI2021MIX+RECOHI2021MIX+HARVESTHI2021PPRECO)
- [1000.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT)
- [10042.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/10042.0_ZMM_13+2017+ZMM_13TeV_TuneCUETP8M1_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano)
- [10024.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/10024.0_TTbar_13+2017+TTbar_13TeV_TuneCUETP8M1_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano)
- [25202.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25+NANOUP15_PU25)
- [10824.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/10824.0_TTbar_13+2018+TTbar_13TeV_TuneCUETP8M1_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano)
- [11634.914](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/11634.914_TTbar_14TeV+2021_DDDDB+TTbar_14TeV_TuneCP5_GenSim+Digi+RecoNano+HARVESTNano+ALCA)
- [11634.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/11634.0_TTbar_14TeV+2021+TTbar_14TeV_TuneCP5_GenSim+Digi+RecoNano+HARVESTNano+ALCA)
- [11634.7](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/11634.7_TTbar_14TeV+2021_trackingMkFit+TTbar_14TeV_TuneCP5_GenSim+Digi+RecoNano+HARVESTNano)
- [12434.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/12434.0_TTbar_14TeV+2023+TTbar_14TeV_TuneCP5_GenSim+Digi+RecoNano+HARVESTNano+ALCA)
- [11634.911](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/11634.911_TTbar_14TeV+2021_DD4hep+TTbar_14TeV_TuneCP5_GenSim+Digi+RecoNano+HARVESTNano+ALCA)
- [10224.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/10224.0_TTbar_13+2017PU+TTbar_13TeV_TuneCUETP8M1_GenSim+DigiPU+RecoFakeHLTPU+HARVESTFakeHLTPU+Nano)
- [39500.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/39500.0_CloseByPGun_CE_H_Coarse_Scint+2026D88+CE_H_Coarse_Scint_GenSimHLBeamSpotHGCALCloseBy+DigiTrigger+RecoGlobal+HARVESTGlobal)
- [39496.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/39496.0_CloseByPGun_CE_E_Front_120um+2026D88+CE_E_Front_120um_GenSimHLBeamSpotHGCALCloseBy+DigiTrigger+RecoGlobal+HARVESTGlobal)
- [11834.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/11834.0_TTbar_14TeV+2021PU+TTbar_14TeV_TuneCP5_GenSim+DigiPU+RecoNanoPU+HARVESTNanoPU)
- [23234.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/23234.0_TTbar_14TeV+2026D49+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal)
- [35034.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/35034.0_TTbar_14TeV+2026D77+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal)
- [28234.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/28234.0_TTbar_14TeV+2026D60+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal)
- [38634.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/38634.0_TTbar_14TeV+2026D86+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal)
- [39434.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/39434.0_TTbar_14TeV+2026D88+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal)
- [250202.181](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/250202.181_TTbar_13UP18+TTbar_13UP18+PREMIXUP18_PU25+DIGIPRMXLOCALUP18_PU25+RECOPRMXUP18_PU25+HARVESTUP18_PU25)
- [35234.999](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrix-results/35234.999_TTbar_14TeV+2026D77PU_PMXS1S2PR+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+PREMIX_PremixHLBeamSpot14PU+DigiTriggerPU+RecoGlobalPU+HARVESTGlobalPU)
</details>



## RelVals-INPUT

- [4.6](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrixINPUT-results/4.6_MinimumBias2010A+MinimumBias2010A+RECOSKIMALCA+HARVESTDR1)```4.6_MinimumBias2010A+MinimumBias2010A+RECOSKIMALCA+HARVESTDR1/step2_MinimumBias2010A+MinimumBias2010A+RECOSKIMALCA+HARVESTDR1.log```
- [136.72411](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrixINPUT-results/136.72411_RunJetHT2016B_reminiaodUL+RunJetHT2016B_reminiaodUL+REMINIAOD_data2016UL_HIPM+HARVESTDR2_REMINIAOD_data2016UL_HIPM)```136.72411_RunJetHT2016B_reminiaodUL+RunJetHT2016B_reminiaodUL+REMINIAOD_data2016UL_HIPM+HARVESTDR2_REMINIAOD_data2016UL_HIPM/step2_RunJetHT2016B_reminiaodUL+RunJetHT2016B_reminiaodUL+REMINIAOD_data2016UL_HIPM+HARVESTDR2_REMINIAOD_data2016UL_HIPM.log```
- [136.72412](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrixINPUT-results/136.72412_RunJetHT2016B_reminiaodUL+RunJetHT2016B_reminiaodUL+REMININANO_data2016UL_HIPM+HARVESTDR2_REMININANO_data2016UL_HIPM)```136.72412_RunJetHT2016B_reminiaodUL+RunJetHT2016B_reminiaodUL+REMININANO_data2016UL_HIPM+HARVESTDR2_REMININANO_data2016UL_HIPM/step2_RunJetHT2016B_reminiaodUL+RunJetHT2016B_reminiaodUL+REMININANO_data2016UL_HIPM+HARVESTDR2_REMININANO_data2016UL_HIPM.log```
<details>
<summary>Expand to see more relval errors ...</summary>

- [136.7611](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrixINPUT-results/136.7611_RunJetHT2016E_reminiaod+RunJetHT2016E_reminiaod+REMINIAOD_data2016_HIPM+HARVESTDR2_REMINIAOD_data2016_HIPM)
- [136.76111](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrixINPUT-results/136.76111_RunJetHT2016E_reminiaodUL+RunJetHT2016E_reminiaodUL+REMINIAOD_data2016UL_HIPM+HARVESTDR2_REMINIAOD_data2016UL_HIPM)
- [136.7721](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrixINPUT-results/136.7721_RunJetHT2016H_reminiaod+RunJetHT2016H_reminiaod+REMINIAOD_data2016+HARVESTDR2_REMINIAOD_data2016)
- [136.77211](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrixINPUT-results/136.77211_RunJetHT2016H_reminiaodUL+RunJetHT2016H_reminiaodUL+REMINIAOD_data2016UL+HARVESTDR2_REMINIAOD_data2016UL)
- [136.8311](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrixINPUT-results/136.8311_RunJetHT2017F_reminiaod+RunJetHT2017F_reminiaod+REMINIAOD_data2017+HARVEST2017_REMINIAOD_data2017)
- [136.83111](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrixINPUT-results/136.83111_RunJetHT2017F_reminiaodUL+RunJetHT2017F_reminiaodUL+REMINIAOD_data2017UL+HARVEST2017_REMINIAOD_data2017UL)
- [136.88811](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrixINPUT-results/136.88811_RunJetHT2018D_reminiaodUL+RunJetHT2018D_reminiaodUL+REMINIAOD_data2018UL+HARVEST2018_REMINIAOD_data2018UL)
- [140.5611](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrixINPUT-results/140.5611_RunHI2018AOD+RunHI2018AOD+REMINIAODHID18+HARVESTHI18MINIAOD)
- [158.01](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrixINPUT-results/158.01_HydjetQ_reminiaodPbPb2018_INPUT+HydjetQ_reminiaodPbPb2018_INPUT+REMINIAODHI2018PPRECOMB+HARVESTHI2018PPRECOMINIAOD)
- [1325.5](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrixINPUT-results/1325.5_ProdZEE_13_reminiaodINPUT+ProdZEE_13_reminiaodINPUT+REMINIAOD_mc2016+HARVESTDR2_REMINIAOD_mc2016)
- [1325.51](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrixINPUT-results/1325.51_TTbar_13_94XreminiaodINPUT+TTbar_13_94XreminiaodINPUT+REMINIAOD_mc2017+HARVESTUP17_REMINIAOD_mc2017)
- [1325.5161](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrixINPUT-results/1325.5161_TTbar_13_reminiaod2016UL_postVFP_INPUT+TTbar_13_reminiaod2016UL_postVFP_INPUT+REMINIAOD_mc2016UL_postVFP+HARVESTDR2_REMINIAOD_mc2016UL_postVFP)
- [1325.516](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrixINPUT-results/1325.516_TTbar_13_reminiaod2016UL_preVFP_INPUT+TTbar_13_reminiaod2016UL_preVFP_INPUT+REMINIAOD_mc2016UL_preVFP+HARVESTDR2_REMINIAOD_mc2016UL_preVFP)
- [1325.518](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrixINPUT-results/1325.518_TTbar_13_reminiaod2018UL_INPUT+TTbar_13_reminiaod2018UL_INPUT+REMINIAOD_mc2018UL+HARVESTUP18_REMINIAOD_mc2018UL)
- [1325.517](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrixINPUT-results/1325.517_TTbar_13_reminiaod2017UL_INPUT+TTbar_13_reminiaod2017UL_INPUT+REMINIAOD_mc2017UL+HARVESTUP17_REMINIAOD_mc2017UL)
- [134.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrixINPUT-results/134.0_MinimumBiasBS+RunMinBias2011B+RECODR1+ALCAPROMPT+ALCAHARVD)
- [144.6](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrixINPUT-results/144.6_MinimumBias2010A+MinimumBias2010A+RECOSKIMALCAR1)
- [1002.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/runTheMatrixINPUT-results/1002.0_RRD+RunMinBias2011A+RECODR1+COPYPASTE)
</details>



## AddOn Tests

- [pat1](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22439/addOnTests/pat1)
```
----- Begin Fatal Exception 15-Feb-2022 18:18:20 CET-----------------------
An exception of category 'ConfigFileReadError' occurred while
   [0] Processing the python configuration file named /data/cmsbld/jenkins/workspace/ib-run-pr-addon/CMSSW_12_3_X_2022-02-14-2300/src/PhysicsTools/PatAlgos/test/IntegrationTest_cfg.py
Exception Message:
 unknown python problem occurred.
RuntimeError: An exception of category 'FileInPathError' occurred.
Exception Message:
edm::FileInPath unable to find file RecoMuon/MuonIdentification/data/mvaID.onnx anywhere in the search path.
The search path is defined by: CMSSW_SEARCH_PATH
${CMSSW_SEARCH_PATH} is: /cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/36963/22439/CMSSW_12_3_X_2022-02-14-2300/poison:/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/36963/22439/CMSSW_12_3_X_2022-02-14-2300/src:/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/36963/22439/CMSSW_12_3_X_2022-02-14-2300/external/slc7_amd64_gcc10/data:/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/poison:/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/src:/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/external/slc7_amd64_gcc10/data
Current directory is: /data/cmsbld/jenkins/workspace/ib-run-pr-addon/addOnTests/pat1


At:
  /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/python/FWCore/ParameterSet/Types.py(840): insertInto
  /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/python/FWCore/ParameterSet/Mixins.py(377): insertContentsInto
  /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/python/FWCore/ParameterSet/Mixins.py(512): insertInto
  /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/python/FWCore/ParameterSet/Modules.py(161): insertInto
  /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/python/FWCore/ParameterSet/Config.py(1145): _insertManyInto
  /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw-patch/CMSSW_12_3_X_2022-02-14-2300/python/FWCore/ParameterSet/Config.py(1400): fillProcessDesc
  <string>(2): <module>

----- End Fatal Exception -------------------------------------------------
```


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 24 differences found in the comparisons
 - DQMHistoTests: Total files compared: 4
 - DQMHistoTests: Total histograms compared: 19811
 - DQMHistoTests: Total failures: 2458
 - DQMHistoTests: Total nulls: 54
 - DQMHistoTests: Total successes: 17299
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 3 files compared)
 - Checked 12 log files, 9 edm output root files, 4 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_12_3_X_2022-02-14-2300+a42efb/48393/triggerResults): no differences found

jpata commented:
@cmsbuild please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/22715/summary.html
**COMMIT**: e2a6cf6997a7986d66060059dde2136c1cb5be9b
**CMSSW**: CMSSW_12_3_X_2022-02-28-1100/slc7_amd64_gcc10
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/36963/22715/install.sh` to create a dev area with all the needed externals and cmssw changes.


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 6 differences found in the comparisons
 - Reco comparison had  1 failed jobs
 - DQMHistoTests: Total files compared: 4
 - DQMHistoTests: Total histograms compared: 19811
 - DQMHistoTests: Total failures: 1476
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 18335
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 3 files compared)
 - Checked 12 log files, 9 edm output root files, 4 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_12_3_X_2022-02-28-1100+a42efb/48610/triggerResults): no differences found


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 4001207
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4001183
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 204 log files, 45 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_3_X_2022-02-28-1100+a42efb/48613/triggerResults): no differences found

jpata commented:
+reconstruction
- technical, no reco-related changes (new CUDA options for ONNXRuntime)

jpata commented:
could this still enter pre6?

perrotta commented:
> could this still enter pre6?

In principle, no:
- We are starting building pre6 kind of now
- This PR was not requested for pre6 during the last ORP, two days ago
- It is not even fully signed

Then, if there are really convinced that this is needed for 12_3_X, please justify it here and we can even decide to delay that build by half a day (and please try also to convince also @cms-sw/heterogeneous-l2 to sign)

makortel commented:
@cmsbuild, please test for slc7_aarch64_gcc11

makortel commented:
I'd prefer to have the use of `CUDAService` in `ONNXRuntime::defaultSessionOptions()` be done in this PR.

But maybe at this stage it would be safer to merge this after 12_3_0_pre6, and backport to 12_3_X after a few successful IBs (including at least one build with ARM and PPC)?

perrotta commented:
> I'd prefer to have the use of `CUDAService` in `ONNXRuntime::defaultSessionOptions()` be done in this PR.
> 
> But maybe at this stage it would be safer to merge this after 12_3_0_pre6, and backport to 12_3_X after a few successful IBs (including at least one build with ARM and PPC)?

I think it is a fair proposal.
Backport of this PR can certainly be approved, once merged and tested in the new master.

jpata commented:
no need to delay the build of course, I was just wondering

cmsbuild commented:
-1

**Failed Tests**:   Build 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/23044/summary.html
**COMMIT**: e2a6cf6997a7986d66060059dde2136c1cb5be9b
**CMSSW**: CMSSW_12_3_X_2022-03-09-2300/slc7_aarch64_gcc11
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/36963/23044/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>>> Entering Package PhysicsTools/ONNXRuntime
>> Leaving Package PhysicsTools/ONNXRuntime
>> Package PhysicsTools/ONNXRuntime built
Entering library rule at PhysicsTools/ONNXRuntime
>> Compiling  /data/cmsbuild/jenkins_b/workspace/ib-run-pr-tests/CMSSW_12_3_X_2022-03-09-2300/src/PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc
/data/cmsbuild/jenkins_b/workspace/ib-run-pr-tests/CMSSW_12_3_X_2022-03-09-2300/src/PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc:19:10: fatal error: cuda_runtime.h: No such file or directory
   19 | #include <cuda_runtime.h>
      |          ^~~~~~~~~~~~~~~~
compilation terminated.
gmake: *** [tmp/slc7_aarch64_gcc11/src/PhysicsTools/ONNXRuntime/src/PhysicsToolsONNXRuntime/ONNXRuntime.cc.o] Error 1
>> Building  shared library tmp/slc7_aarch64_gcc11/src/PhysicsTools/ONNXRuntime/src/PhysicsToolsONNXRuntime/libPhysicsToolsONNXRuntime.so
</pre>

hqucms commented:
> I'd prefer to have the use of `CUDAService` in `ONNXRuntime::defaultSessionOptions()` be done in this PR.
> 
> But maybe at this stage it would be safer to merge this after 12_3_0_pre6, and backport to 12_3_X after a few successful IBs (including at least one build with ARM and PPC)?

Sorry for the delay on this PR, I was occupied by a series urgent things this periods...

@makortel I thought a bit about the use of `CUDAService` -- I feel it would probably be better to use that in the EDModules, while keeping only a minimal dependency of ONNXRuntime package on other CMSSW packages -- so the ONNXRuntime package only depends on the external `onnxruntime` and `FWCore`. I think that allows a better separation of the packages and leaves more flexibility for various use cases of ONNXRuntime (so it could be used even outside EDModules as well if needed). 

We can even remove the CUDA availability detection in ONNXRuntime, so it purely gets the flag from outside (and using CUDAService in the EDModules).

What do you think?

makortel commented:
@hqucms In general I'm in favor of keeping dependencies limited. I guess this could imply e.g. removing the `auto_gpu` option from the `GPUMode` enumeration, i..e the caller fully specifies the behavior.

It would make the calling code (EDModule) a bit more complex, but maybe that's ok. In principle one could abstract the checking logic to a helper function, but then we'd have a question of where to place that helper function if not here.

About

> (so it could be used even outside EDModules as well if needed)

I just want to note that the dependence on MessageLogger makes that a bit more difficult (for which an easiest, but not necessarily prettiest, solution would be to have `#ifdef` for the standalone case; probably best to be left at the time when such use is needed).

The ARM test shows that the use of CUDA Runtime API doesn't work (transitively) when ONNX wasn't compiled with CUDA support. From the discussion above those calls would be removed anyway. I wonder if `AppendExecutionProvider_CUDA()` is available at compile time in such case?

vlimant commented:
The code overhead for the calling module looks minimal to me, and hence we should remove the detection from within ONNXRuntime ; if that's the show-stopper for this to move forward.

vlimant commented:
The code overhead for the calling module looks minimal to me, and hence we should remove the detection from within ONNXRuntime ; if that's the show-stopper for this to move forward.

hqucms commented:
Sorry for the long delay on this. I have implemented the proposed changes and removed all the CUDA detection part.

> The ARM test shows that the use of CUDA Runtime API doesn't work (transitively) when ONNX wasn't compiled with CUDA support. From the discussion above those calls would be removed anyway. I wonder if AppendExecutionProvider_CUDA() is available at compile time in such case?

I think it's available at compile time, but let's test it to be sure.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36963/28965

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #36963 was updated. @makortel, @slava77, @clacaputo, @cmsbuild, @fwyzard, @jpata can you please check and sign again.

makortel commented:
@cmsbuild, please test for slc7_aarch64_gcc11

makortel commented on PhysicsTools/ONNXRuntime/interface/ONNXRuntime.h:
```diff
@@ -29,6 +29,8 @@ namespace cms::Ort {
     ONNXRuntime& operator=(const ONNXRuntime&) = delete;
     ~ONNXRuntime();
 
+    static ::Ort::SessionOptions defaultSessionOptions(bool use_cuda = false);
```
Maybe an enumeration (`enum class`) would be more self-documenting and extensible?

hqucms replied:
@makortel So something like:

```c++
enum class OrtSessOpt {
  cpu,
  cuda,
};
```

makortel replied:
Yeah, something like that.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/23327/summary.html
**COMMIT**: 8930cc43d18544f687defb4fba73b896aebcd724
**CMSSW**: CMSSW_12_4_X_2022-03-23-1100/slc7_amd64_gcc10
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/36963/23327/install.sh` to create a dev area with all the needed externals and cmssw changes.


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 24 differences found in the comparisons
 - DQMHistoTests: Total files compared: 4
 - DQMHistoTests: Total histograms compared: 19874
 - DQMHistoTests: Total failures: 1936
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 17938
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 3 files compared)
 - Checked 12 log files, 9 edm output root files, 4 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_12_4_X_2022-03-23-1100+a42efb/49109/triggerResults): no differences found


## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-a42efb/39434.911_TTbar_14TeV+2026D88_DD4hep+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3695650
 - DQMHistoTests: Total failures: 51219
 - DQMHistoTests: Total nulls: 65
 - DQMHistoTests: Total successes: 3644344
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 204 log files, 45 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_4_X_2022-03-23-1100+a42efb/49114/triggerResults): no differences found

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/23335/summary.html
**COMMIT**: 8930cc43d18544f687defb4fba73b896aebcd724
**CMSSW**: CMSSW_12_4_X_2022-03-22-2300/slc7_aarch64_gcc11
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/36963/23335/install.sh` to create a dev area with all the needed externals and cmssw changes.

jpata commented:
@cmsbuild please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/23815/summary.html
**COMMIT**: 8930cc43d18544f687defb4fba73b896aebcd724
**CMSSW**: CMSSW_12_4_X_2022-04-11-1100/slc7_amd64_gcc10
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/36963/23815/install.sh` to create a dev area with all the needed externals and cmssw changes.


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 24 differences found in the comparisons
 - DQMHistoTests: Total files compared: 4
 - DQMHistoTests: Total histograms compared: 19874
 - DQMHistoTests: Total failures: 955
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 18919
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 3 files compared)
 - Checked 12 log files, 9 edm output root files, 4 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_12_4_X_2022-04-11-1100+a42efb/49533/triggerResults): no differences found


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3593039
 - DQMHistoTests: Total failures: 8
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3593009
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 200 log files, 45 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_4_X_2022-04-11-1100+a42efb/49536/triggerResults): no differences found

jpata commented:
Should we aim to have this in the prerelease next week? Is there still something pending here?
Also pinging the ML-prod team @riga @yongbinfeng on the timeline of this feature.

fwyzard commented:
> There is my suggestion to use `enum class` instead of `bool` for denoting where to run [#36963 (comment)](https://github.com/cms-sw/cmssw/pull/36963#discussion_r833800258), that I think would be easier to address now rather than later.

-heterogeneous

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36963/29454

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #36963 was updated. @makortel, @slava77, @clacaputo, @cmsbuild, @fwyzard, @jpata can you please check and sign again.

hqucms commented:
@makortel The `enum class` is used implemented in https://github.com/cms-sw/cmssw/pull/36963/commits/5eb31fbe030dc87b4e19fa4501d44b3b7d5c3560.
Sorry for the long delay, this somehow slipped through the cracks...

tvami commented:
@cmsbuild , please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a42efb/24133/summary.html
**COMMIT**: 5eb31fbe030dc87b4e19fa4501d44b3b7d5c3560
**CMSSW**: CMSSW_12_4_X_2022-04-22-1100/slc7_amd64_gcc10
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/36963/24133/install.sh` to create a dev area with all the needed externals and cmssw changes.


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 24 differences found in the comparisons
 - DQMHistoTests: Total files compared: 4
 - DQMHistoTests: Total histograms compared: 19874
 - DQMHistoTests: Total failures: 1195
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 18679
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 3 files compared)
 - Checked 12 log files, 9 edm output root files, 4 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_12_4_X_2022-04-22-1100+a42efb/49782/triggerResults): no differences found


## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-a42efb/39434.75_TTbar_14TeV+2026D88_HLT75e33+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HLT75e33+HARVESTGlobal

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3695434
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3695410
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 205 log files, 45 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_4_X_2022-04-22-1100+a42efb/49785/triggerResults): no differences found

makortel commented:
+1

Thanks @hqucms

jpata commented:
+reconstruction
- technical, readiness for CUDA use in ONNXRuntime
- no changes

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

qliphy commented:
+1
